chips/lowrisc/src/hmac.rs

```diff
@@ -224,37 +258,60 @@ impl<'a> hil::digest::DigestData<'a, 32> for Hmac<'a> {
     fn add_data(
         &self,
         data: LeasableBuffer<'static, u8>,
-    ) -> Result<usize, (ErrorCode, &'static mut [u8])> {
-        let regs = self.registers;
-
-        regs.cmd.modify(CMD::START::SET);
+    ) -> Result<(), (ErrorCode, LeasableBuffer<'static, u8>)> {
+        if self.busy.get() {
+            Err((ErrorCode::BUSY, data))
+        } else {
+            self.busy.set(true);
+            self.data.set(Some(LeasableBufferDynamic::Immutable(data)));
+
+            let regs = self.registers;
+            regs.cmd.modify(CMD::START::SET);
+            // Clear the FIFO empty interrupt
+            regs.intr_state.modify(INTR_STATE::FIFO_EMPTY::SET);
+            // Enable interrupts
+            regs.intr_enable.modify(INTR_ENABLE::FIFO_EMPTY::SET);
+            let ret = self.data_progress();
 
-        // Clear the FIFO empty interrupt
-        regs.intr_state.modify(INTR_STATE::FIFO_EMPTY::SET);
+            if ret {
+                regs.intr_test.modify(INTR_TEST::FIFO_EMPTY::SET);
+            }
 
-        // Enable interrupts
-        regs.intr_enable.modify(INTR_ENABLE::FIFO_EMPTY::SET);
+            Ok(())
+        }
+    }
 
-        // Set the length and data index of the data to write
-        self.data_len.set(data.len());
-        self.data.set(Some(data));
-        self.data_index.set(0);
+    fn add_mut_data(
+        &self,
+        data: LeasableMutableBuffer<'static, u8>,
+    ) -> Result<(), (ErrorCode, LeasableMutableBuffer<'static, u8>)> {
+        if self.busy.get() {
+            Err((ErrorCode::BUSY, data))
+        } else {
+            self.busy.set(true);
+            self.data.set(Some(LeasableBufferDynamic::Mutable(data)));
+
+            let regs = self.registers;
+            regs.cmd.modify(CMD::START::SET);
+            // Clear the FIFO empty interrupt
+            regs.intr_state.modify(INTR_STATE::FIFO_EMPTY::SET);
+            // Enable interrupts
+            regs.intr_enable.modify(INTR_ENABLE::FIFO_EMPTY::SET);
+            let ret = self.data_progress();
 
-        // Call the process function, this will start an async fill method
-        let ret = self.data_progress();
+            if ret {
+                regs.intr_test.modify(INTR_TEST::FIFO_EMPTY::SET);
```

Comment:
It is correct, it's a little odd though.

We need to do this if we can an `add_data()` call that does NOT fill up the buffer. If we write all of the data in one go it's possible we won't get a FIFO empty interrupt (image writing a single byte). We could use deferred callbacks, but this feature already exists in the hardware so it seemed easier

Reply:
Shouldn't there be an interrupt even if the FIFO isn't filled? I.e., the interrupt is EMPTY, not NOT_FULL. A single byte write should push an entry into the FIFO, which then falls idle. If you aren't getting an interrupt, this seems like a hardware bug/issue we should reach out to the OT team about.

The problem with this approach is that you're going to handle interrupts at the interrupt handling rate, rather than for each FIFO. E.g., it's possible you get an interrupt but there's actually no space because it was fast.

Reply:
This came up on the call today. The lowrisc HMAC engine is fast enough (for SHA) that it's quite possible that it can process faster than the memory bus can feed it data. This means that `lowrisc::Hmac::process` might just loop for the entire passed slice.